doc/api-management.md

```diff
@@ -0,0 +1,118 @@
+# API Management
+
+> Guidance below for Theia framework. End products will get better development experience by using private visibility.
+
+- [**Stability**](#stability)
+  - [**Experimental**](#experimental)
+  - [**Stable**](#stable)
+- [**Finalization**](#finalization)
+- [**Deprecation**](#deprecation)
+
+Theia is a framework embracing openness, extensibility, and customizability as much as possible:
+- API defaults to public for clients.
+- API defaults to protected for extenders.
+- Language constructions prohibiting runtime access to internals are never used and so on.
+
+Usually, the version management is built around API visibility.
+Particularly, if a public API is broken then the major release is required.
+
+Following the conventional approach is not practicable for Theia since all APIs are more or less public.
+It will slow down API innovation, accumulate technical debt or require many major releases.
+
+Instead, we require a major release only if a stable API is broken.
+Breaking an experimental API can be done with a minor release.
+
+## Stability
+
+Conceptually, an API is all assumptions adopters depend on to make use or extend Theia.
+A stable API is based on assumptions that are not subject to change.
+It does not matter whether an API is public or internal, or whether it is used a lot or never.
+
+> For instance, the language server protocol depends on such data types like URIs and positions
+since they don't change from language to language. It ensures its stability.
+
+API stability is indicated explicitly by adding `@experimental` or `@stable` js-doc tags.
+An API without a stability tag is considered to be experimental.
+
+The explicit stability tag should mention since which version an API is stable or experimental.
+
+```ts
+/**
+ * @since 0.1.0 - added in 0.1.0
```

Comment:
The `since` tag is not mentioned in the text. Would you propose to transform `experimental` into `since` upon finalization?

Reply:
> The since tag is not mentioned in the text.

I've just wanted to be explicit here. But if it is obvious then I can remove it.

The thing is I don't want to bother anyone with annotations while they adding new APIs except the case when they touch stable APIs already. Otherwise it increase friction on PR reviews which is already quite high.

> Would you propose to transform experimental into since upon finalization?

Not sure about transforming but adding since tag to stable API sounds like a good idea. I will add it.

Reply:
Another pro for since tag is linting. If we want to have something like PDE api tools it would help instead of using git and github API, but we can add it later too.